scripts/scrape-data-from-api.js

```diff
@@ -0,0 +1,151 @@
+/**
+ * As originally designed, the mobilev.is website served JSON-formatted data
+ * from an HTTP API that was backed by a PostgreSQL database. It also relied on
+ * images hosted in an Amazon Web Services S3 bucket.
+ *
+ * This script was written to retrieve the JSON data and the images in the
+ * interest of simplifying the site's architecture. The files it creates may be
+ * checked in to the project's repository, so that (following some minor
+ * alterations to the front-end application, the site may be deployed as a
+ * collection of static files.
+ */
+'use strict';
+
+const {createWriteStream} = require('fs');
+const fs = require('fs').promises;
+const http = require('http');
+const https = require('https');
+const os = require('os');
+const path = require('path');
+
+const limit = (max, fn) => {
+  const queued = [];
+  let count = 0;
+
+  return function () {
+    let go;
+    const wrapped = new Promise((resolve, reject) => {
+      go = () => {
+        count += 1;
+        return fn.apply(this, arguments)
+          .then(resolve, reject)
+          .then(() => {
+            count -= 1;
+            if (queued.length) {
+              queued.shift()();
+            }
+          });
+      };
+    });
+
+    count < max ? go() : queued.push(go);
+
+    return wrapped;
+  };
+};
+
+const fetchJSON = limit(5, (url) => {
+  return new Promise((resolve, reject) => {
+    const request = http.get(url, (response) => {
+      if (response.statusCode !== 200) {
+        reject(new Error(`Unexpected status code (${response.statusCode}) for response from ${url}`));
+        response.resume();
+        return;
+      }
+
+      let body = '';
+      response.setEncoding('utf8');
+      response.on('data', (chunk) => body += chunk);
+      response.on('end', () => {
+        try {
+          resolve(JSON.parse(body));
+        } catch (error) {
+          reject(Error(`Unable to parse JSON at ${url}: ${error}`));
+        }
+      });
+    });
+
+    request.on('error', reject);
+  });
+});
+
+const mkDirP = (filePath) => {
+  const mkdir = (dirName) => fs.mkdir(dirName)
+    .catch((err) => { if (err.code !== 'EEXIST') { throw err; } });
+
+  return filePath
+    .split(path.sep)
+    .map((_, index, parts) => path.join(...parts.slice(0, index + 1)))
+    .reduce(
+      (promise, dirName) => promise.then(() => mkdir(dirName)),
+      Promise.resolve()
+    );
+};
+
+const fetchImageWithCache = async (domain, id) => {
+  const fileName = path.join('data', 'images', id);
```

Comment:
If there is a convenient library or tool we can use I think we should try to determine the image mime type from the image header and add the appropriate extension. Browsers will determine that on their own but without the extension most servers will report the mime type `application/binary-stream` or `octet/binary-stream`.

Reply:
That's a good idea, though it does mean modifying the data that we store on disk. Just a little extra effort.

On the bright side, Amazon's serving at least some of this content with a Content-Type header, so we might not have to resort to content sniffing:

    $ curl --head https://s3.amazonaws.com/mobilevis.bocoup.com/images/d9611091-0e4b-45a2-9610-48a8e9d7a5c8
    HTTP/1.1 200 OK
    x-amz-id-2: 22E73JwxWP2EhwOPeVNaKY13twuvxqOUoSLr651uJ15ibq8RaaymNgB8uboaj//+zNCetBkGPIA=
    x-amz-request-id: 02832FE9CPYRXT4P
    Date: Fri, 11 Mar 2022 01:11:48 GMT
    Last-Modified: Wed, 15 Jun 2016 20:41:41 GMT
    ETag: "3a260bb5d3c8cc184db867e61f17f335-1"
    Accept-Ranges: bytes
    Content-Type: image/png
    Server: AmazonS3
    Content-Length: 98769